util/util.go

```diff
@@ -102,3 +106,58 @@ func GetOrCreateOwnLog(t tserv.Threadservice, id thread.ID) (info thread.LogInfo
 	err = t.Store().AddLog(id, info)
 	return info, err
 }
+
+// AddPeerFromAddress parses the given address and adds the dialable component
+// to the peerstore.
+// If a dht is provided and the address does not contain a dialable component,
+// it will be queried for peer info.
+func AddPeerFromAddress(addrStr string, pstore peerstore.Peerstore) (pid peer.ID, err error) {
+	addr, err := ma.NewMultiaddr(addrStr)
+	if err != nil {
+		return
+	}
+	p2p, err := addr.ValueForProtocol(ma.P_P2P)
+	if err != nil {
+		return
+	}
+	pid, err = peer.IDB58Decode(p2p)
+	if err != nil {
+		return
+	}
+	dialable, err := GetDialable(addr)
+	if err == nil {
+		pstore.AddAddr(pid, dialable, peerstore.PermanentAddrTTL)
+	}
+
+	return pid, nil
+}
+
+// GetDialable returns the portion of an address suitable for storage in a peerstore.
+func GetDialable(addr ma.Multiaddr) (ma.Multiaddr, error) {
+	parts := strings.Split(addr.String(), "/"+ma.ProtocolWithCode(ma.P_P2P).Name)
+	return ma.NewMultiaddr(parts[0])
+}
+
+// CanDial returns whether or not the address is dialable.
+func CanDial(addr ma.Multiaddr, s *swarm.Swarm) bool {
+	parts := strings.Split(addr.String(), "/"+ma.ProtocolWithCode(ma.P_P2P).Name)
+	addr, _ = ma.NewMultiaddr(parts[0])
+	tr := s.TransportForDialing(addr)
+	return tr != nil && tr.CanDial(addr)
+}
+
+// DecodeKey from a string into a symmetric key.
+func DecodeKey(k string) (*sym.Key, error) {
+	b, err := base58.Decode(k)
+	if err != nil {
+		return nil, err
+	}
+	return sym.NewKey(b)
+}
+
+// PadArgs returns args with new length pad.
+func PadArgs(args []string, len int) []string {
+	padded := make([]string, len)
+	copy(padded, args)
```

Comment:
Nitpick: copy will copy min of `len` and `len(args)`. If `len(args)` greater than `len`, something bad would happen. Maybe is at the risk of the caller, just mentioning it before seeing where is used.

Reply:
The caller _should_ have the sense that after doing something like `args = PadArgs(args, 2)`, they cannot access `args[3]`. Maybe a better name would be `TruncateArgs`? Or maybe `padded := make([]string, max(len, len(args))`

Reply:
Oh, `TruncateArgs` feels much better seeing this code. The other option changes what the function does. I'll leave it to whatever you think is the best :)